File: leetcode/explore/binary tree/binary tree preorder traversal.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                result.append(node.val)
                stack.append(node.right)
                stack.append(node.left)
        
        return result


if __name__ == "__main__":
    input = TreeNode(1)
    input.right = TreeNode(2)
    cur= input.right
    cur.left = TreeNode(3)
    cur = cur.left
    print(Solution().preorderTraversal(input))

File: leetcode/explore/binary tree/binary tree preorder traversal.cpp
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> result;
        stack<TreeNode*> nodeStack;
        nodeStack.push(root);
        while (!nodeStack.empty())
        {
            auto node = nodeStack.top();
            nodeStack.pop();
            if (node != NULL)
            {
                result.push_back(node->val);
                nodeStack.push(node->right);
                nodeStack.push(node->left);
            }
        }

        return result;
    }
};


int main() {
    TreeNode* root = new TreeNode(1);
    TreeNode* cur = root;
    cur->right = new TreeNode(2);
    cur = cur->right;
    cur->left = new TreeNode(3);
    Solution s;
    s.preorderTraversal(root);

    return 0;
}